Outliers detection

hhan14 opened the pull request:
This is updated version, which is debugged and run-able.
But I`d like to improve it with more functional specifications as the
current version just meets conceptual purpose of the issue.

Please feel free to add comments, thoughts, and suggestions for further enhancement.

han-health-ai committed:
Initial checking for outliers

Not yet work

han-health-ai committed:
This is updataed version, which is debugged and runnable.
But I`d like to improve it with more functional specifications as the
current version just meets conceptual purpose of the issue.

omidkj commented on hs628_tools/Outlier_debugged.py:
```diff
@@ -0,0 +1,35 @@
+import csv
+import matplotlib.pyplot as plt
+
+##Class Outliers:
```
You may need to uncomment this line to initiate your class Outliers.

hhan14 replied:
Thanks for commenting Omid: I made the change for the work. :)

omidkj commented on hs628_tools/Outlier_debugged.py:
```diff
@@ -0,0 +1,35 @@
+     self.std = std
+     self.outliers = outliers
+
+with open("data.csv") as f:
```
You don't necessarily  need to read your CSV file inside your class. you can read it in main program and pass it as a parameter to the outliers method.

hhan14 replied:
Thanks again, I modified it too for the final work. :)

han-health-ai committed:
Renamed file, improved functions and added doc strings

kamehta2 commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,66 @@
+        Unit tests:
+          >>> "6".isnumeric()
+          True
+          >>> "None".isnumeric()
```
The doctests you are running do not actually test your implementation, but instead they are only verifying the .isnumeric function that you use within your code. I think it will help if you actually add tests that will test your functions.

hhan14 replied:
Hi Kinjal, the unit test doc string is just to confirm the data cleaning purpose. Thanks.

han-health-ai committed:
Removed data read part used inside of init function

han-health-ai committed:
corrected variable names

han-health-ai committed:
Critical changes added:
(1) Improved the second function based on the same basic function but for working on each function
(2) Changed from test data('mydata') to real data set, "data.csv".
(3) Standardized codes and document styles according to 'PEP 8' guidelines for Python

han-health-ai committed:
Added a note that "Assumed the data set('data.csv') is loaded."

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        (3) Convert values from string to float
+
+        Unit tests to confirm data cleaning:
+          >>> "6".isnumeric()
```
I'm glad to see some doctests, but these particular ones are actually just testing whether or not the ```string.isnumeric()``` method is working (i.e., whether the script is running under python 3), not whether your own method is working correctly (which is what they really should be testing).

hhan14 replied:
I agree with your point and erased those doctests samples in the final version.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+    outliers
+
+    *Note:
+    Assumed the data set('data.csv') is loaded
```
I'm not understanding, what is it about the code that assumes this particular data file is 'loaded'?

hhan14 replied:
I removed the note and added outside function, "read_testdata",  to read the data set.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        self.data = data
+
+        if data is None:
+            for i in range(len(self.data)):  # remove None in data.csv
```
What we had talked about, which I see in one of the older commits, is only reading the default datafile if the provided data was None.  From the commit message it looks clear you intentionally removed it, but then you need to also either remove or do something else with the rest of this code-- if data was None then this code would cause an Exception.  And if a non-None data object were provided, then the elements would probably not be all strings, that was really just a result of how you had been reading your default data.

hhan14 replied:
This code has been removed. The data file is read in the outside function as mentioned above comment.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        """
+        Actions:
+        (0) Define parameters
+        (1) Load data
```
action 1 "Load Data" is no longer an action taken here, and I'm doubtful actions 2 & 3 should be here either (see below)

hhan14 replied:
All the documentation in this section is removed.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+
+        if data is None:
+            for i in range(len(self.data)):  # remove None in data.csv
+                self.data[i] = [elem for elem in self.data[i] if elem.isnumeric()]
```
this list interpretation would drop elements (columns) if any are not numeric, which could easily lead to rows having a different number of elements, and misaligned columns.  

Also, when we spoke about this check before, I hadn't realized that you were running python 3 and that isnumeric() is not a regular string method in python 2.7; we'd prefer (though it's not mandated) that these modules are all compatible with python 2.7.  A further complication is that isnumeric() doesn't accept strings with decimal points, so if you were going to pursue this kind of check you'd probably be best off writing your own function for it.

It's fine to instead just require that the user pass numeric data without any missing values (just note that in the docstring).

hhan14 replied:
I added the Python version 3.6 was used to write on the top of the project. Also, removed this code and replaced with new functional codes by converting string values to float in the "def read_testdata' function part.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        out_lim = np.selp.std*1.96  # set outlier limit
+        outliers = np.ones((my_arr.shape[0],), dtype=np.bool)
+        self.mean = np.self.mean(my_arr, axis=0)
+        self.std = np.self.std(my_arr, axis=0, drop=1)
```
These two lines look like you were trying to decide between using the ```self.data```,   ```self.mean``` and ```self.std``` from the init, versus just using newly provided my_array, and wound up sort of combining them, but it won't work like this.   Either way could be OK, but not this mishmash.  If you wanted to be more like sklearn then you could provide the std dev limit in the init, and then provide the data here.  Also I don't find numpy.std as having a 'drop' parameter for python 2.7 nor for python 3.6

hhan14 replied:
"np.self.mean" is replaced with "np.mean", and the same procedure is applied to the other line for "self.std"

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+            self.data = np.array(self.data, dtype=np.float)  # convert to float
+            print(self.data)
+    ##
+    def cal_outliers(my_arr, out_lim):
```
- A method should always have ```self``` as the first parameter.  
- For me at least, "cal" is not an intuitive abbreviation for "calculate"-- if you want to abbreviate it please use "calc".
- It's great to have a docstring; please don't forget to describe the parameters-- summarizing the function and describing the parameters and any return value are the most important aspects of the docstring.  I comment below on the summary.  It's also not clear to me why ```my_arr``` should be a parameter here, and I think that ```out_lim``` would be better with a fuller, more descriptive name.

hhan14 replied:
I added "self" as the first parameter, and changed "cal_outliers" to "calc_outliers". I also added and modified description of parameters, and documentation of the function.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+    def cal_outliers(my_arr, out_lim):
+        """
+        Concept:
+        Function for calculating values out of scope of two standard deviation in data and returning those values as
```
This summary, and some of the current code, treats the cutoff as fixed at 2 stdev, but then you have the ```out_lim``` parameter....  Also, this first use of "values" only seems accurate if I interpret "values" as "samples" (since it looks like you return a collection of rows) but then that interpretation doesn't really  work for the second use of "values".  Also please say more clearly just what the method returns.

hhan14 replied:
I changed terms of "values" and "samples", and clarified the sentence.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+                 Proceedings of the Thirtieth International Florida Artificial Intelligence Research Society Conference)
+        """
+        my_arr = np.random.randn(20, 10)  # data array
+        out_lim = np.selp.std*1.96  # set outlier limit
```
these first two lines appear left over from testing, and would prevent the method from working for any provided data.

hhan14 replied:
This codes are taken off.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        self.std = np.self.std(my_arr, axis=0, drop=1)
+        for i in range(my_arr.shape[1]):  # added for detect outliers by each column
+            col = my_arr[:, i]
+            outliers[outliers] &= np.abs((col[outliers] - self.mean[i]) / self.std[i]) < out_lim
```
It looks to me like this loop & calculation are actually winnowing down those rows that do ___NOT___ contain any outlier values, which very well might be the best approach, but then the ```outliers``` variable is misnamed.

hhan14 replied:
This part of code is fixed.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        for i in range(my_arr.shape[1]):  # added for detect outliers by each column
+            col = my_arr[:, i]
+            outliers[outliers] &= np.abs((col[outliers] - self.mean[i]) / self.std[i]) < out_lim
+        return my_arr[outliers]
```
As it stands , it looks to me like this will be returning those rows that do not contain any outlier values.  That could be fine, but just think through the most likely use-case for your user, return an appropriate value for that, and make sure that your function is named appropriately and that the docstring matches.  It seems it would probably be easy enough to make this aspect be user-controllable-- whether it is the outlier rows, or non-outlier-rows, that get returned.

hhan14 replied:
I agree with what you commented, and like the idea of user-controllable one. I initially wanted not only to get collection of outliers specific location of the outliers but also to detect specific locations of those outliers with the row and column information. I still would like to enhance the current version to meet my initial goal.

EHWUSF commented on hs628_tools/Outliers.py:
```diff
@@ -0,0 +1,76 @@
+        outliers
+
+        Actions:
+        Get quartile values for outlier limits: out of two standard deviation
```
I'm not seeing how this really corresponds to quartile values.

hhan14 replied:
This is erased.

EHWUSF commented:
For Issues, it is common to assign them to the person taking on primary responsibility for the implementation.  For PRs, if they are assigned to someone it is to indicate that person is taking (or being assigned) primary responsibility for reviewing the PR and for making the decision about merging it (so a PR is virtually never assigned to the person who submitted it).

han-health-ai committed:
Made changes upon the review comments

han-health-ai committed:
Fixed function to read_testdata

han-health-ai committed:
Debugged functions

han-health-ai committed:
This is the final version of project:
- Documentations are revised for each function, variable, and note
- Data set is replaced with "data.csv" from test data set, "mydata.csv"

han-health-ai committed:
This is the final version polished all documentations and commented comments.
